app/services/upcoming_offline_reservation_notifier.rb

```diff
@@ -0,0 +1,28 @@
+class UpcomingOfflineReservationNotifier
+
+  def notify
+    upcoming_offline_reservations.each do |reservation|
+      UpcomingOfflineReservationMailer.generate_mail(reservation).deliver_later
+    end
+  end
+
+  private
+
+  def offline_instrument_ids
+    Instrument.where(schedule_id: offline_schedule_ids)
+  end
+
+  def offline_schedule_ids
+    OfflineReservation.current.joins(:product).pluck(:schedule_id)
+  end
+
+  def upcoming_offline_reservations
+    Reservation
```

Comment:
I feel like I've seen a very similar chain of scopes somewhere else. Would it be worth rolling them up in `Reservation` into a new scope?